std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -196,259 +26,132 @@ private static Pattern buildPattern(
+
+      // Do we need to rescan?
+      if (!isInMixedState && parser.numberWithSeparators() != NumberWithSeparators.DOT_COMMA) {
+        builder =
+            makeBuilderWithCapacity(sourceStorage.size(), problemAggregator.createSimpleChild());
+        for (int j = 0; j < i; j++) {
+          var subText = sourceStorage.getItemBoxed(j);
+          var subResult = subText == null ? null : parseSingleValue(subText, problemAggregator);
+          if (subResult == null) {
+            builder.appendNulls(1);
+          } else {
+            builder.append(subResult);
+          }
+        }
       }
 
-      index++;
-      pattern = patternForIndex(index);
-    }
-
-    CommonParseProblemAggregator aggregator = problemAggregator.createContextAwareChild();
-    Builder fallback = makeBuilderWithCapacity(sourceStorage.size(), aggregator);
-    parseColumnWithPattern(patternForIndex(bestIndex), sourceStorage, fallback, aggregator);
-    return fallback.seal();
-  }
-
-  private int parseColumnWithPattern(
-      Pattern pattern,
-      Storage<String> sourceStorage,
-      Builder builder,
-      ParseProblemAggregator aggregator) {
-    Context context = Context.getCurrent();
-    for (int i = 0; i < sourceStorage.size(); i++) {
-      var text = sourceStorage.getItemBoxed(i);
-      if (text == null) {
+      // Append the result
+      if (result == null) {
         builder.appendNulls(1);
       } else {
-        var value = innerParseSingleValue(text, pattern);
-        if (value != null) {
-          builder.appendNoGrow(value);
-        } else {
-          if (aggregator == null) {
-            return i;
-          }
-
-          aggregator.reportInvalidFormat(text);
-          builder.appendNulls(1);
-        }
+        builder.append(result);
       }
 
       context.safepoint();
     }
-    return -1;
-  }
 
-  @Override
-  protected Builder makeBuilderWithCapacity(int capacity, ProblemAggregator problemAggregator) {
-    return allowDecimal
-        ? NumericBuilder.createDoubleBuilder(capacity, problemAggregator)
-        : NumericBuilder.createLongBuilder(capacity, integerTargetType, problemAggregator);
+    return builder.seal();
   }
 
-  private Object innerParseSingleValue(String text, Pattern pattern) {
-    if (allowDecimal) {
-      var trimmed = trimValues ? text.trim() : text;
-      if (trimmed.equals("NaN")) {
-        return Double.NaN;
-      }
-      if (trimmed.equals("Infinity")) {
-        return Double.POSITIVE_INFINITY;
-      }
-      if (trimmed.equals("-Infinity")) {
-        return Double.NEGATIVE_INFINITY;
-      }
-    }
+  @Override
+  public Object parseSingleValue(String text, ParseProblemAggregator problemAggregator) {
+    var result = parser.parse(text, isInteger);
 
-    var parsed = pattern.matcher(text);
-    if (!parsed.matches()) {
+    if (result instanceof FormatDetectingNumberParser.NumberParseFailure) {
+      problemAggregator.reportInvalidFormat(text);
```

Comment:
Probably would be nice to propagate the `message` from `NumberParseFailure` in some way.

Not necessarily for now, but I'd add a TODO.

Reply:
Added to #11228.